### ISSUE TITLE
When to worry about Imprecise conversion / Max diff

### ISSUE THREAD

**Issue by chadrockey** [AlexanderLutsenko/nobuco#60]
Occasionally I see some indications that a conversion shows up with the Inaccurate warning.  When googling I can’t find any other guidance on these numbers for similar tools.  I’m sure it’s ultimately domain specific, but do you have guidelines on when these imprecisions can happen and when to worry about them?

Yellow — conversion imprecise

For instance, on a few Conv2d[torch.nn.modules.conv], I see (!) Max diff 0.00281 (0.144%)

Is there anything to watch out for that would cause these?  Does nobuco track if the errors could potentially compound?

I see some debug output about CUDA from Tensorflow immediately before the conversion.  Is this conversion done on CPU or GPU?  If both are possible, is there a preference?

**Comment by AlexanderLutsenko** [comment]
Hi!
Imagine these are the outputs of the original Pytorch op/module, and its Keras counterpart:

```python
output_pt = [100000.0, 100000.0, 100000.0]
output_tf = [100000.0, 100000.0, 100000.1]
```
Then, the max absolute and relative diffs are

```python
max_abs_diff = abs(100000.0 - 100000.1) = 0.1
max_rel_diff = (100000.1 / 100000.0 - 1) * 100 = 0.0001%
```
and we get this warning:

```console
(!) Max diff 0.1 (0.0001%)
```

Although `max_abs_diff` is high, that's _likely_ not a reason to worry because `max_rel_diff` is tiny. Thing is, Nobuco can never know it for sure as it's very task-specific. 

Take a look at this example:

```python
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(128, 256, kernel_size=(3, 3))

    def forward(self, x):
        x = self.conv(x)
        x = nn.ReLU()(x)
        return x


x = torch.normal(0, 1, size=(1, 128, 32, 32))
pytorch_module = MyModule().eval()
```

![1](https://github.com/user-attachments/assets/9edd3b8c-4bed-4021-b3d9-0aa691901b76)

All green and beautiful. How can we destroy the harmony? Turns out, convolutions work a bit differently in the two frameworks. We can multiply the input by, say, 100 to amplify the difference.

```python
    def forward(self, x):
        x = x * 100
        x = self.conv(x)
        x = nn.ReLU()(x)
        return x
```

![2](https://github.com/user-attachments/assets/33703122-011b-4cd0-9ddc-06a820944fcc)

There are two things to note here
1. The `ReLU` is still green. 
How come? The `Conv2D` produces a slightly different output, which is the `ReLU`'s input, so that `ReLU` should output slightly different values, too. Right? Yes, but that way, you could believe there is something wrong with `ReLU` converter, which is not true. To eliminate this source of confusion, each converted module is fed the original Pytorch inputs, therefore tested in isolation.
2. The `Conv2D`'s output discrepancy is propagated to its outer module, `MyModule`.

Now, let's replace `ReLU` with some bounded function, like `Sigmoid`.

```python
    def forward(self, x):
        x = x * 100
        x = self.conv(x)
        x = nn.Sigmoid()(x)
        return x
```

![3](https://github.com/user-attachments/assets/18ccbc87-adc6-4ee1-90e9-ad98d0ceea52)

`MyModule` is back to green. Being bounded, that `Sigmoid` brings its input values closer together thereby mitigating the output differences. When you see the outermost module green, you probably don't need to worry even if some inner modules are painted yellow.

**Comment by chadrockey** [comment]
Great, thanks for clarifying this.  I assumed it was very important given that you put this much effort to colorize and run these checks.  I couldn't find any guides for when to be concerned or to establish an initial understanding of the error levels and causes.
